weasyl/test/test_login.py

```diff
@@ -0,0 +1,477 @@
+# This Python file uses the following encoding: utf-8
+import unittest
+import bcrypt
+import arrow
+import web
+import ast
+
+import db_utils
+from weasyl import login, macro
+from weasyl import define as d
+from weasyl.error import WeasylError
+
+# Main test account
+raw_password = "0123456789"
+bcrypt_hash = login.passhash(raw_password)
+old_2a_bcrypt_hash = "$2a$12$qReI924/8pAsoHu6aRTX2ejyujAZ/9FiOOtrjczBIwf8wqXAJ22N."
+
+class Bag(object):
+    def __init__(self, **kw):
+        for kv in kw.items():
+            setattr(self, *kv)
+
+class SigninTestCase(unittest.TestCase):
+    def testVerifyLoginRecordIsUpdated(self):
+        user_id = db_utils.create_user(username='testloginsuite')
+        d.execute("UPDATE login SET last_login = '-1' WHERE userid = %i", [user_id])
+        # login.signin(user_id) -will- raise an AttributeError when d.web.ctx.weasyl_session
+        #   tries to execute itself; so catch/handle
+        try:
+            login.signin(user_id)
+        except AttributeError:
+            print 'Caught AttributeError, as expected'
+        query = d.execute("SELECT last_login FROM login WHERE userid = %i", [user_id])
+        self.assertGreater(query[0][0], -1)
+        d.execute("DELETE FROM login WHERE userid = %i", [user_id])
+
+class AuthenticateBcryptTestCase(unittest.TestCase):
+    def testNoUsernameProvided(self):
+        result = login.authenticate_bcrypt(username='', password='password')
+        self.assertTrue(result[0] == 0)
+        self.assertTrue(result[1] == 'invalid')
+
+    def testNoPasswordProvided(self):
+        result = login.authenticate_bcrypt(username='testloginsuite', password='')
+        self.assertTrue(result[0] == 0)
+        self.assertTrue(result[1] == 'invalid')
+
+    def testInvalidUsernameProvided(self):
+        result = login.authenticate_bcrypt(username='InvalidUsername', password=raw_password)
+        self.assertTrue(result[0] == 0)
+        self.assertTrue(result[1] == 'invalid')
+
+    def testLoginFailsForModsWithInvalidAuthentication(self):
+        log_path = '%s%s.%s.log' % (macro.MACRO_SYS_LOG_PATH, 'login.fail', d.get_timestamp())
+        open(log_path, 'w').close()
+        mod_userid = 2061
+        user_id = db_utils.create_user(username='ikani')
+        d.execute("UPDATE login SET userid = %i WHERE userid = %i", [mod_userid, user_id])
```

Comment:
I think this is the part where being able to update the staff entries by passing in a new dict would be really useful.